synapse/handlers/federation_event.py

```diff
@@ -1515,10 +1515,9 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase) -> No
         # support it or the event is not from the room creator.
         room_version = await self._store.get_room_version(marker_event.room_id)
         create_event = await self._store.get_create_event_for_room(marker_event.room_id)
-        room_creator = create_event.content.get(EventContentFields.ROOM_CREATOR)
         if not room_version.msc2716_historical and (
             not self._config.experimental.msc2716_enabled
-            or marker_event.sender != room_creator
+            or marker_event.sender != create_event.sender
```

Comment:
Is the existing logic correct? Shouldn't it be `if not room_version.msc2716_historical OR (...)`?

(`A => B` iff `not A or B`.)

Additionally: I think (strictly speaking) that this makes MSC2716 depend on MSC2175?

Reply:
> Is the existing logic correct? Shouldn't it be `if not room_version.msc2716_historical OR (...)`?

I'm not sure, would need to ask @MadLittleMods.

> Additionally: I think (strictly speaking) that this makes MSC2716 depend on MSC2175?

Yes, I guess this should probably use the same or logic that we use in other spots.

Reply:
(I fixed this to break the dependency. 🤷 )

Reply:
I think the logic correct. This is taking a lot of squinting so it could probably be clearer but here is the breakdown:

 - If the room version supports MSC2716: we always process marker events (no early return)
 - If the room version does not support MSC2716 (existing room versions), the MSC2716 experimental feature has to be enabled and the event has to be from the room creator
 
If we used `OR` here, then the second case with existing room versions would never be work.

---

https://github.com/matrix-org/synapse/blob/b682443d2414658f56f30030c406fb6a05659f08/synapse/handlers/federation_event.py#L1522-L1524

If we break these down into conditions, then the current boolean logic looks like: `not V and (not E or not C)`
```py
is_supported_room_version = room_version.msc2716_historical # V
is_msc2716_enabled = self._config.experimental.msc2716_enabled # E
is_event_from_room_creator = marker_event.sender == room_creator # C
```

If we add some labels and derive a positive `should_process_event` condition which we can then later negate, then the boolean logic looks like `not (V or (E and C))` which simplifies to the exact same logic above (see alternate form listed on [Wolfram Alpha](https://www.wolframalpha.com/input?i=not+%28V+or+%28E+and+C%29%29))

This probably does make more sense though:
```py
is_event_elgible_to_process_in_existing_room_versions = is_msc2716_enabled and is_event_from_room_creator
should_process_event = is_supported_room_version or is_event_elgible_to_process_in_existing_room_versions
if not should_process_event:
    return
```

Reply:
Aha, thanks for double checking. (I'd missed that this logic was guarding an early return, amongst other things.)